editor/editor_properties_array_dict.cpp

```diff
@@ -321,100 +357,56 @@ void EditorPropertyArray::update_property() {
-
-			if (value_type == Variant::NIL && subtype != Variant::NIL) {
-				value_type = subtype;
-			}
+		for (Slot &slot : slots) {
+			int idx = slot.index;
+			Variant::Type value_type = subtype;
 
-			if (value_type == Variant::OBJECT && Object::cast_to<EncodedObjectAsID>(value)) {
-				EditorPropertyObjectID *editor = memnew(EditorPropertyObjectID);
-				editor->setup("Object");
-				prop = editor;
-			} else {
-				prop = EditorInspector::instantiate_property_editor(nullptr, value_type, "", subtype_hint, subtype_hint_string, PROPERTY_USAGE_NONE);
+			if (value_type == Variant::NIL) {
+				value_type = array.get(idx).get_type();
 			}
 
-			prop->set_object_and_property(object.ptr(), prop_name);
-			prop->set_label(itos(i + offset));
-			prop->set_selectable(false);
-			prop->set_use_folding(is_using_folding());
-			prop->connect("property_changed", callable_mp(this, &EditorPropertyArray::_property_changed));
-			prop->connect("object_id_selected", callable_mp(this, &EditorPropertyArray::_object_id_selected));
-			prop->set_h_size_flags(SIZE_EXPAND_FILL);
-			prop->set_read_only(is_read_only());
-			hbox->add_child(prop);
-
-			bool is_untyped_array = array.get_type() == Variant::ARRAY && subtype == Variant::NIL;
-
-			if (is_untyped_array) {
-				Button *edit_btn = memnew(Button);
-				edit_btn->set_icon(get_editor_theme_icon(SNAME("Edit")));
-				hbox->add_child(edit_btn);
-				edit_btn->set_disabled(is_read_only());
-				edit_btn->connect("pressed", callable_mp(this, &EditorPropertyArray::_change_type).bind(edit_btn, i + offset));
-			} else {
-				Button *remove_btn = memnew(Button);
-				remove_btn->set_icon(get_editor_theme_icon(SNAME("Remove")));
-				remove_btn->set_disabled(is_read_only());
-				remove_btn->connect("pressed", callable_mp(this, &EditorPropertyArray::_remove_pressed).bind(i + offset));
-				hbox->add_child(remove_btn);
+			// Check if the editor property needs to be updated.
+			bool value_as_ID = Object::cast_to<EncodedObjectAsID>(array.get(idx));
+			if (value_type != slot.type || (value_type == Variant::OBJECT && (value_as_ID ^ slot.as_ID))) {
+				slot.type = value_type;
+				EditorProperty *new_prop = nullptr;
+				if (value_type == Variant::OBJECT && value_as_ID) {
+					EditorPropertyObjectID *editor = memnew(EditorPropertyObjectID);
+					editor->setup("Object");
+					new_prop = editor;
+				} else {
+					new_prop = EditorInspector::instantiate_property_editor(nullptr, value_type, "", subtype_hint, subtype_hint_string, PROPERTY_USAGE_NONE);
+				}
+				new_prop->set_selectable(false);
+				new_prop->set_use_folding(is_using_folding());
+				new_prop->connect("property_changed", callable_mp(this, &EditorPropertyArray::_property_changed));
+				new_prop->connect("object_id_selected", callable_mp(this, &EditorPropertyArray::_object_id_selected));
+				new_prop->set_h_size_flags(SIZE_EXPAND_FILL);
+				new_prop->set_read_only(is_read_only());
+				slot.prop->add_sibling(new_prop);
+				slot.prop->queue_free();
```

Comment:
```suggestion
				memfree(slot.prop);
```
Unless there is a need to deffer freeing, `memfree()` should be preferred.